src/lqcontrol.jl

```diff
@@ -107,81 +83,103 @@ function update_values!(lq::LQ)
+    A0, B0 = sqrt(lq.bet) * A, sqrt(lq.bet) * B
+    P = solve_discrete_riccati(A0, B0, R, Q, N)
 
     # Compute F
-    S1 = Q .+ lq.bet .* (B' * P * B)
-    S2 = lq.bet .* (B' * P * A)
-    F = S1 \ S2
+    s1 = Q + lq.bet * (B' * P * B)
+    s2 = lq.bet * (B' * P * A) + N
+    F = s1 \ s2
 
     # Compute d
-    d = lq.bet .* trace(P * C * C') / (1 - lq.bet)
+    d = lq.bet * trace(P * C * C') / (1 - lq.bet)
 
     # Bind states
     lq.P, lq.F, lq.d = P, F, d
-    nothing
 end
 
 function stationary_values(lq::LQ)
-    stationary_values!(lq)
-    return lq.P, lq.F, lq.d
+	_lq = LQ(
+			copy(lq.Q), 
+			copy(lq.R), 
+			copy(lq.A), 
+			copy(lq.B), 
+			copy(lq.C), 
+			copy(lq.N),
+			copy(lq.bet), 
+			lq.term,
+			copy(lq.rf))
+
+    stationary_values!(_lq)
+    return _lq.P, _lq.F, _lq.d
 end
 
+# dispatch for a scalar problem
+function _compute_sequence{T}(lq::LQ, x0::T, policies)
+	term = length(policies)
+	x_path = Array(T, term+1)
+	u_path = Array(T, term)
+	
+	x_path[1] = x0
+	u_path[1] = -(first(policies)*x0)
+	w_path    = lq.C * randn(term+1)
+
+    for t = 2:term
+        f = policies[t]
+        x_path[t] = lq.A*x_path[t-1] + lq.B*u_path[t-1] + w_path[t]
+        u_path[t] = -(f*x_path[t])
+    end
+    x_path[end] = lq.A*x_path[term] + lq.B*u_path[term] + w_path[end]
 
-function compute_sequence(lq::LQ, x0::ScalarOrArray, ts_length=100)
-    # simplify notation
-    Q, R, A, B, C = lq.Q, lq.R, lq.A, lq.B, lq.C
+    x_path, u_path, w_path
+end
+
+# dispatch for a vector problem
+function _compute_sequence{T}(lq::LQ, x0::Vector{T}, policies)
+	n, j, term = length(x0), length(lq.C), length(policies)
+	x_path = Array(T, n, term+1)
+	u_path = Array(T, n, term)
+	
+	x_path[:,1] = x0
+	u_path[:,1] = -(first(policies)*x0)
+	w_path      = [dot(lq.C,randn(j)) for i=1:(term+1)]
+
+    for t = 2:term
+        f = policies[t]
+        x_path[:,t] = lq.A*x_path[t-1] + lq.B*u_path[t-1] + w_path[t]
+        u_path[:,t] = -(f*x_path[t])
+    end
+    x_path[:,end] = lq.A*x_path[:,term] + lq.B*u_path[:,term] + w_path[end]
+
+    x_path, u_path, w_path
+end
 
-    # Preliminaries,
-    if lq.T != nothing
+function compute_sequence(lq::LQ, x0::ScalarOrArray, ts_length=100)
+    if lq.term != nothing
         # finite horizon case
-        T = min(ts_length, lq.T)
-        lq.P, lq.d = lq.Rf, 0.0
+        term = min(ts_length, lq.term)
+        lq.P, lq.d = lq.rf, 0.0
     else
         # infinite horizon case
-        T = ts_length
+        term = ts_length
         stationary_values!(lq)
     end
 
-    # Set up initial condition and arrays to store paths
-    x0 = reshape([x0], lq.n, 1)  # make sure x0 is a column vector
-    x_path = Array(eltype(x0), lq.n, T+1)
-    u_path = Array(eltype(x0), lq.k, T)
-    w_path = C * randn(lq.j, T+1)
-
     # Compute and record the sequence of policies
-    policies = Array(typeof(lq.F), T)
-
-    for t=1:T
-        if lq.T != nothing
+    policies = Array(typeof(lq.F), term)
+    for t = 1:term
```

Comment:
I know this was the original structure, but I think we can do better here.

How about we do something like:

```
if lq.term != nothing
    policies = Array(typeof(lq.F), term)
    for t=1:term
        update_values!(lq)
        policies[t] = lq.F
    end
else
    policies = fill(lq.F, term)
end
```

I don't see a need to check  `lq.term != nothing` each iteration.